addOns/ascanrulesAlpha/src/main/java/org/zaproxy/zap/extension/ascanrulesAlpha/XSLTInjection.java

```diff
@@ -0,0 +1,260 @@
+        "<xsl:variable name=\"rtobject\" select=\"runtime:getRuntime()\"/>\n"
+                + "<xsl:variable name=\"process\" select=\"runtime:exec($rtobject,'erroneous_command')\"/>\n"
+                + "<xsl:variable name=\"waiting\" select=\"process:waitFor($process)\"/>\n"
+                + "<xsl:value-of select=\"$process\"/>",
+        "<xsl:value-of select=\"php:function('exec','erroneous_command 2>&amp;1')\"/>"
+    };
+
+    private static final String[] evidenceCmdExec = {
+        "Cannot run program", "erroneous_command: not found",
+    };
+
+    private enum XSLTInjectionType {
+        ERROR(errorCausingPayloads, evidenceError, "error"),
+        VENDOR(vendorReturningPayloads, evidenceVendors, "vendor"),
+        PORTSCAN(new String[] {}, evidencePortScanning, "portscan"),
+        COMMAND_EXEC(cmdExecPayloads, evidenceCmdExec, "command");
+
+        private String[] payloads;
+        private String[] responses;
+        private String resourceIdentifier;
+
+        XSLTInjectionType(String[] payloads, String[] responses, String resourceIdentifier) {
+            this.payloads = payloads;
+            this.responses = responses;
+            this.resourceIdentifier = resourceIdentifier;
+        }
+    }
+
+    private static final Logger LOG = Logger.getLogger(XSLTInjection.class);
+
+    // used to check against the attack strength
+    private int requestsSent = 0;
+    // map setting the limits associated with each attack strength
+    private static Map<AttackStrength, Integer> strengthToRequestCountMap = new HashMap<>();
+
+    static {
+        strengthToRequestCountMap.put(AttackStrength.LOW, 6);
+        strengthToRequestCountMap.put(AttackStrength.MEDIUM, 12);
+        strengthToRequestCountMap.put(AttackStrength.HIGH, 24);
+    }
+
+    @Override
+    public void scan(HttpMessage msg, String param, String value) {
+        // sets the payloads that recquire an instance to be determined
+        XSLTInjectionType.PORTSCAN.payloads = new String[] {getXslForPortScan()};
+
+        // goes through all checks and stops if it finds a possible
+        // injection
+        for (XSLTInjectionType check : XSLTInjectionType.values()) {
+            if (tryInjection(msg, param, check)) {
+                return;
+            }
+        }
+    }
+
+    private Boolean tryInjection(HttpMessage msg, String param, XSLTInjectionType checkType) {
+        for (String payload : checkType.payloads) {
+            try {
+                if (isStop() || requestsLimitReached()) { // stop before sending request
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Scanner " + getName() + " Stopping.");
+                    }
+                    return false;
+                }
+                msg = sendRequest(msg, param, payload);
+
+                for (String response : checkType.responses) {
+                    if (getBaseMsg().getResponseBody().toString().contains(response)) {
```

Comment:
Ideally we should exclude these responses before sending any message, we also just need to check once per base message.

Reply:
I'm not too sure about if I misunderstood this comment, but I changed the check in the newest commit (alongside other fixes), now it checks before request if any of the responses are present to skip.

Reply:
It can be done before the first for loop and it should filter/collect instead. Then in this for it will just check the ones that were not present in the message.